les/fetcher.go

```diff
@@ -65,42 +65,49 @@ type response struct {
 
 // fetcherPeer holds the fetcher-specific information for each active peer
 type fetcherPeer struct {
-	latest       *announceData          // The latest announcement sent from the peer
-	trustedMap   map[common.Hash]uint64 // Trusted announces map
-	trustedQueue *prque.Prque           // Trusted announces queue
+	latest *announceData // The latest announcement sent from the peer
+
+	// These following two fields can track the latest announces
+	// from the peer with limited size for caching.
+	announces     map[common.Hash]*announce // Announcement map
+	announceQueue *prque.Prque              // Announcement queue
 }
 
-// addTrustedAnno enqueues an new trusted announcement. If the queued
-// announces overflow, evict from the oldest.
-func (fp *fetcherPeer) addTrustedAnno(number uint64, hash common.Hash) {
+// addAnno enqueues an new trusted announcement. If the queued announces overflow,
+// evict from the oldest.
+func (fp *fetcherPeer) addAnno(announce *announce) {
 	// Short circuit if the announce already exists. In normal case it should
 	// never happen since only monotonic announce is accepted. But the adversary
 	// may feed us fake announces with higher td but same hash. In this case,
 	// ignore the announce anyway.
-	if _, exist := fp.trustedMap[hash]; exist {
+	hash, number := announce.data.Hash, announce.data.Number
+	if _, exist := fp.announces[hash]; exist {
 		return
 	}
-	fp.trustedMap[hash] = number
-	fp.trustedQueue.Push(hash, -int64(number))
+	fp.announces[hash] = announce
+	fp.announceQueue.Push(hash, -int64(number))
 
 	// Evict oldest if the announces are oversized.
-	for fp.trustedQueue.Size() > trustedItemsThreshold {
-		item, _ := fp.trustedQueue.Pop()
-		delete(fp.trustedMap, item.(common.Hash))
+	for fp.announceQueue.Size() > cachedAnnosThreshold {
+		item, _ := fp.announceQueue.Pop()
+		delete(fp.announces, item.(common.Hash))
 	}
 }
 
-// forwardTrustedAnno removes all announces from the map with a number lower than
+// forwardAnno removes all announces from the map with a number lower than
```

Comment:
I think this announcement queue design has a problem. You put them into a queue ordered by block number. Here you also evict old ones based on block number. I think you should instead order and evict based on Td because it is the Td, not the number that should always grow monotonically (actually you don't even need an ordered queue structure because monotonic Td is already enforced in the announcements). So you can just put the incoming announcements in a FIFO list and compare the oldest element to the Td of the current chain. A new head with a same or lower block number might still be important if it has a high Td but anything that has an equal or lower Td than the current chain is not interesting any more so it is safe to throw away.

Reply:
Fixed